File: wildcardPatternMatching.js
/* Given a text and a wildcard pattern, implement wildcard pattern matching algorithm that finds if wildcard 
pattern is matched with text. The matching should cover the entire text (not partial text).

The wildcard pattern can include the characters ‘?’ and ‘*’
‘?’ – matches any single character
‘*’ – Matches any sequence of characters (including the empty sequence)

For example,

Text = "baaabab",
Pattern = “*****ba*****ab", output : true
Pattern = "baaa?ab", output : true
Pattern = "ba*a?", output : true
Pattern = "a*ab", output : false 

Each occurrence of ‘?’ character in wildcard pattern can be replaced with any other character and each occurrence 
of ‘*’ with a sequence of characters such that the wildcard pattern becomes identical to the input string after 
replacement.
 */

/* Pseudocode:
- Have a helper function to get rid of duplicate asterisks from pattern, call it eliminateDuplicateAsterisk
   - This function takes in the pattern input string called patternStr as a parameter
   - Declare an empty string called newPattern to hold the duplicate * free new pattern string
   - loop through pattern starting w/ for loop
      - check if current character and next character is '*' then just do nothing & continue
      - otherwise, add the current character to the newPattern string 
   - Outside for loop, return the newly updated string newPattern

- Have a helper function called getSubstrTillNextStar:
   - This helper function will take in an input string call it inpStr and pattern string call it patternStr,
      -It'll start at index 0 and get the substring till next star, store it in a variable call subString
      - To do this start at index 0 of the patternStr, and in while loop see that the character in patternStr at patternIndex 
      is not equal to '*' then store the character in patternStr at patternIndex in subString and then increment patternIndex
      - Next outside while loop, find the subString in the inpStr input string by using indexOf method and store that index
      in a variable called nextIndex

- Have a pointer to input string and a pointer to pattern string call them strPointer & patternPointer, respectively
- Have a while loop of while strPointer is less than length of input string str and patternPointer is less than input pattern string,
pattern
   - Have another while loop of while character in pattern at patternPointer is equal to * (pattern[patternPointer] === '*') 
   increment patternPointer
   - After patternPointer has been incremented if there was a * at that index in pattern and reached a non * character then check if
   the character in pattern at patternPointer index is equal to ? (pattern[patternPointer] === '?') then increment patternPointer
   by 1 and increment strPointer by 1
   
 */

let eliminateDuplicateAsterisk = (patternStr) => {
   let newPattern = "";
   for (let i = 0; i < patternStr.length; i++) {
     if (patternStr[i] === "*" && patternStr[i + 1] === "*") {
       continue;
     } else {
       newPattern += patternStr[i];
     }
   }
   return newPattern;
 };
 
//  console.log(eliminateDuplicateAsterisk("*****ba*****ab")); // "*ba*ab"
//  console.log(eliminateDuplicateAsterisk("ba*a?")); // "ba*a?"
//  console.log(eliminateDuplicateAsterisk("*a*b")); // "*a*b"
//  console.log(eliminateDuplicateAsterisk("*")); // "*"

let strrmatch = (str, pattern) => {
   let newPattern = eliminateDuplicateAsterisk(pattern);
   // console.log('newPattern.length:', newPattern.length)
   let strPointer = 0;
   let patternPointer = 0;
   let finalRes = false;

   while (strPointer < str.length && patternPointer < newPattern.length - 1) {
      // console.log('strPointer: ', strPointer);
      // console.log('patternPointer: ', patternPointer);
      while (newPattern[patternPointer] === '*') {
         patternPointer++;
      };

      while (newPattern[patternPointer] === '?') {
         patternPointer++;
         strPointer++;
      }

      if (str[strPointer] === newPattern[patternPointer]) {
         finalRes = true;
         strPointer++;
         patternPointer++;
         // return true;
      }
      else {
         finalRes = false;
      }
   }

   return finalRes;
}

console.log(strrmatch("baaabab", "*****ba*****ab")); // true  *ba*ab
console.log(strrmatch("baaabab", "baaa?ab")); // true
console.log(strrmatch("baaabab", "ba*a?")); // true
console.log(strrmatch("baaabab", "a*ab")); // false
console.log(strrmatch("aa", "a")); // false
console.log(strrmatch("aa", "*")); // true
console.log(strrmatch("cb", "?a")); // false
console.log(strrmatch("adceb", "*a*b")); // true
console.log(strrmatch("acdcb", "a*c?b")); // false

File: wildcardPatternMatching.py
'''
Given a text and a wildcard pattern, implement wildcard pattern matching algorithm that finds if wildcard 
pattern is matched with text. The matching should cover the entire text (not partial text).

The wildcard pattern can include the characters ‘?’ and ‘*’
‘?’ – matches any single character
‘*’ – Matches any sequence of characters (including the empty sequence)

For example,

Text = "baaabab",
Pattern = “*****ba*****ab", output : true
Pattern = "baaa?ab", output : true
Pattern = "ba*a?", output : true
Pattern = "a*ab", output : false

Each occurrence of ‘?’ character in wildcard pattern can be replaced with any other character and each occurrence 
of ‘*’ with a sequence of characters such that the wildcard pattern becomes identical to the input string after replacement.
'''

def strrmatch(strr, pattern, n, m):
    if (m == 0): 
        return (n == 0) 
          
    # lookup table for storing results of 
    # subproblems 
    lookup = [[False for i in range(m + 1)] for j in range(n + 1)]
    print('lookup: ', lookup)


print(strrmatch("baaabab", "*****ba*****ab", 2, 3))

# print(strrmatch("baaabab", "*****ba*****ab", len('baaabab'), len('*****ba*****ab')))
# print(strrmatch("baaabab", "baaa?ab", 2, 3))
# print(strrmatch("baaabab", "ba*a?", 2, 3))
# print(strrmatch("baaabab", "a*ab", 2, 3))

# Lambda solution from Sean
class Solution:
    def isMatch(self, s: str, p: str)-> bool:
        # use a dynamic programming approach 
        # think of DP as recursion that is augmented with caching 
        cache = {}
        
        def recurse(s, p):
            # first just think of this with plain old recursion
            # what are our base cases? 
            # base case 1: if the string s is empty 
            if not s:
                # s = "", p = ""
                # if the pattern is empty, "*", True
                # s = "", p = " "
                # if the pattern is not empty, "?" False 
                # s = "", p = "*"
                return (p == '') or (p[0] == '*' and recurse(s, p[1:]))
            # base case 2: if the pattern is empty
            if not p:
                # p = "", then s = "", True
                # else, False 
                return (s == '')
            
            # also check the cache 
            if (s, p) not in cache:
                # handling asterisks in pattern 
                # what happens when p[0] == '*'? 
                if p[0] == '*':
                    # match any number of chars
                    # or match the empty sequence 
                    # recurse twice, one way that increments the string 
                    # the other way incrementing the pattern 
                    # if any recursive come back True, return True 
                    cache[(s, p)] = recurse(s, p[1:]) or recurse(s[1:], p)
                # otherwise, p[0] != "*"
                else:
                    # compare the first chars of s and p
                    # if they don't match or p[0] != '?'
                    # return False
                    if s[0] != p[0] and p[0] != '?':
                        cache[(s, p)] = False
                    # if we've recursed all the way through both strings
                    # with no mismatches between string and pattern, return True
                    else:
                        cache[(s, p)] = recurse(s[1:], p[1:])
            
            return cache[(s, p)]
        
        return recurse(s, p)

        # without dynamic programming
        def recurse(s, p):
            # first just think of this with plain old recursion
            # what are our base cases? 
            # base case 1: if the string s is empty 
            if not s:
                # s = "", p = ""
                # if the pattern is empty, "*", True
                # s = "", p = " "
                # if the pattern is not empty, "?" False 
                # s = "", p = "*"
                return p == '' or p[0] == '*' and recurse(s, p[1:])
            # base case 2: if the pattern is empty
            if not p:
                # p = "", then s = "", True
                # else, False 
                return s == ''
            
            # also check the cache 
            # handling asterisks in pattern 
            # what happens when p[0] == '*'? 
            if p[0] == '*':
                # match any number of chars
                # or match the empty sequence 
                # recurse twice, one way that increments the string 
                # the other way incrementing the pattern 
                # if any recursive come back True, return True 
                return recurse(s, p[1:]) or recurse(s[1:], p)
            # otherwise, p[0] != "*"
            else:
                # compare the first chars of s and p
                # if they don't match or p[0] != '?'
                # return False
                if s[0] != p[0] and p[0] != '?':
                    return False
                # if we've recursed all the way through both strings
                # with no mismatches between string and pattern, return True
                return recurse(s[1:], p[1:])
    
        return recurse(s, p)


# Lambda solution from KJ
'''
class Solution:
    def isMatch(self, s: str, p: str) -> bool:
        
        i, j = 0, 0                        # pointers for indices of s and p
        wildcards = {'*', '?'}             # dict to store wildcard characters
        star_index_in_p = -1               # pointer for the last index of '*' in p
        
        while i < len(s):
            
            # if we're beyond pattern's limit, or pattern is an unmatched character and not a wildcard
            if j >= len(p) or (p[j] != s[i] and p[j] not in wildcards):
                if star_index_in_p == -1:  # if we haven't seen a '*' in p yet, we have no flexibility
                    return False
                j = star_index_in_p + 1    # reset p index pointer to character after star_index_in_p
                star_index_in_s += 1       # reset s index pointer to character after star_index_in_s
                i = star_index_in_s        # i.e. use '*' to match one character from s
                
            elif p[j] == '*':              # record index of '*' in p and next index to be matched in s
                star_index_in_p = j
                star_index_in_s = i
                j += 1
                
            else:                          # else, if characters match or p[j]=='?', increment both pointers
                i += 1
                j += 1
                
        while j < len(p):                  # now that we're outside the limits of s,
            if p[j] != '*':                # any remaining characters in p can only be '*'
                return False
            j += 1
            
        return True                        # if we made it this far, s and p match
    
        # Time complexity: O(n*m) -> O(n)
        # Space complexity: O(1)
'''